Add Health Checks for Multi-Site deployment to HA Docs

kami619 opened the pull request:
Fixes https://github.com/keycloak/keycloak/issues/33143

This PR aims to improve the troubleshooting aspects of a Multi-Site deployment of Keycloak on a Kubernetes environment and helps assert the overall health status of various components involved in the deployment.

kami619 committed:
Add Health Checks for Multi-Site guide to Docs

Closes #33143

Signed-off-by: Kamesh Akella <kamesh.asp@gmail.com>

ahus1 reviewed (dismissed):
Thank you for this guide. IMHO it would be sufficient to assume only Kubernetes in this deployment and not OpenShift, which would make it more approachable to the wider community. 

So I'd ask you to change "oc" to "kubectl" and replace "OpenShift" with "Kubernetes" except where we reference the prerequisite of the previous installation explicitly.

kami619 commented:
@ahus1 I have addressed your previous review comments, please take another look if that's enough.

ahus1 committed:
Review

Closes #33143

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (commented):
Thank you for he docs. I updated the formatting a bit and shortened some of the headings and sentences for better readability. Please let me know if that breaks any meaning. 

I'll ask Andy to a review of the guide before this will then be merged.

ahus1 commented:
@andymunro - we would like to include this guide in KC26. Can you please review this new text? Thanks!

ahus1 commented:
@andymunro - gentle reminder. Please review when you have the time. We'd like to merge it tomorrow the latest to finish this task which is in our sprint. If you don't have the time, We could merge this version and postpone a review until the RHBK release. Thanks!

andymunro commented:
Thanks for the reminder, @ahus1.  I will check it now.

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+When running the <@links.ha id="introduction" /> in a Kubernetes environment,
+you should automate checks to see if everything is up and running as expected.
+
+This page provides an overview of URLs,
+Kubernetes resources and Healthcheck endpoints available to verify a multi-site setup of {project_name}.
```
```suggestion
This page provides an overview of URLs,
Kubernetes resources, and Healthcheck endpoints available to verify a multi-site setup of {project_name}.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+== Overview
+
+A proactive monitoring strategy aims to detect and alert on issues before they impact users. It is the key for a highly resilient and highly available {project_name} application.
```
```suggestion
A proactive monitoring strategy aims to detect and alert about issues before they impact users. This strategy is the key for a highly resilient and highly available {project_name} application.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+A proactive monitoring strategy aims to detect and alert on issues before they impact users. It is the key for a highly resilient and highly available {project_name} application.
+
+Health checks across various architectural components (like application health, load balancing, caching, and overall system status) are critical for:
```
```suggestion
Health checks across various architectural components (such as application health, load balancing, caching, and overall system status) are critical for:
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Health checks across various architectural components (like application health, load balancing, caching, and overall system status) are critical for:
+
+Ensuring High Availability:: By verifying that all sites and the load balancer are operational, it helps guarantee that the system can handle requests even if one site goes down.
```
```suggestion
Ensuring High Availability:: Verifying that all sites and the load balancer are operational is a key to ensure that  a system can handle requests even if one site goes down.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+. https://kubernetes.io/docs/tasks/tools/#kubectl[Kubectl CLI is installed and configured].
+
+. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your OS already.
```
```suggestion
. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your operating system already.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Ensuring High Availability:: By verifying that all sites and the load balancer are operational, it helps guarantee that the system can handle requests even if one site goes down.
+
+Maintaining Performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
```
```suggestion
Maintaining performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Maintaining Performance:: Checking the health and distribution of the {jdgserver_name} cache ensures that {project_name} can maintain optimal performance by efficiently handling sessions and other temporary data.
+
+Operational Resilience:: By continuously monitoring the health of both {project_name} and its dependencies within the Kubernetes environment, the system can quickly identify and possibly auto-remediate issues, reducing downtime.
```
```suggestion
Operational resilience:: By continuously monitoring the health of both {project_name} and its dependencies within the Kubernetes environment, the system can quickly identify and possibly auto-remediate issues, reducing downtime.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+. Install https://jqlang.github.io/jq/download/[jq] if it is not available on your OS already.
+
+== Specific Health Checks
```
```suggestion
== Specific health checks
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+== Specific Health Checks
+
+=== {project_name} Load Balancer and Sites
```
```suggestion
=== {project_name} load balancer and sites
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Verifies the health of the {project_name} application through its load balancer and both primary and backup sites. This ensures that {project_name} is accessible and that the load balancing mechanism is functioning correctly across different geographical or network locations.
+
+This command returns the health status of the {project_name} application's connection to its configured database, thus confirming the reliability of db connections.
```
```suggestion
This command returns the health status of the {project_name} application's connection to its configured database, thus confirming the reliability of database connections.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+Verifies the health of the {project_name} application through its load balancer and both primary and backup sites. This ensures that {project_name} is accessible and that the load balancing mechanism is functioning correctly across different geographical or network locations.
+
+This command returns the health status of the {project_name} application's connection to its configured database, thus confirming the reliability of db connections.
+This is only available on the management port, and not available from the external URL.
+In a Kubernetes setup, the sub-status `health/ready` is checked periodically to make the Pod as ready.
```
```suggestion
This command is available only on the management port and not from the external URL.
In a Kubernetes setup, the sub-status `health/ready` is checked periodically to make the Pod as ready.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+curl -s https://keycloak-load-balancer-url/lb-check
+----
+
+These commands will return the running status of the Site A and Site B of the {project_name} in a cross-site setup (be it Active/Passive or Active/Active).
```
```suggestion
These commands will return the running status of the Site A and Site B of the {project_name} in a cross-site setup (whether it is Active/Passive or Active/Active).
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+curl -s https://keycloak_site_b_url/lb-check
+----
+
+=== {jdgserver_name} Cache Health
+Check the health of the default cache manager and individual caches in an external {jdgserver_name} cluster.
+This is vital for {project_name} performance and reliability,
+as {jdgserver_name} is often used for distributed caching and session clustering in {project_name} deployments.
```
```suggestion
=== {jdgserver_name} Cache health
Check the health of the default cache manager and individual caches in an external {jdgserver_name} cluster.
This check is vital for {project_name} performance and reliability,
as {jdgserver_name} is often used for distributed caching and session clustering in {project_name} deployments.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+This is vital for {project_name} performance and reliability,
+as {jdgserver_name} is often used for distributed caching and session clustering in {project_name} deployments.
+
+This command returns the overall health of the {jdgserver_name} cache manager, this is useful as the Admin user doesn't need to provide user credentials to get the health status.
```
```suggestion
This command returns the overall health of the {jdgserver_name} cache manager, which is useful as the Admin user does not need to provide user credentials to get the health status.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+curl -s https://infinispan_rest_url/rest/v2/cache-managers/default/health/status
+----
+
+Whereas for these health checks, Admin user needs to provide the {jdgserver_name} user credentials as part of the request to peek into the overall health of the external {jdgserver_name} cluster caches.
```
```suggestion
In contrast to the preceding health checks, the following health checks require the Admin user to provide the {jdgserver_name} user credentials as part of the request to peek into the overall health of the external {jdgserver_name} cluster caches.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+----
+
+The `jq` filter is a convenience to compute the overall health based on the individual cache health.
+Optionally you can run the above command without the `jq` filter to see the full details.
```
```suggestion
You can also choose to run the above command without the `jq` filter to see the full details.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+The `jq` filter is a convenience to compute the overall health based on the individual cache health.
+Optionally you can run the above command without the `jq` filter to see the full details.
+
+=== {jdgserver_name} Cluster Distribution
```
```suggestion
=== {jdgserver_name} Cluster distribution
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+ | jq --argjson expectedCount 3 'if map(select(.node_addresses | length > 0)) | length == $expectedCount then "HEALTHY" else "UNHEALTHY" end'
+----
+
+=== Overall, {jdgserver_name} System Health
```
```suggestion
=== Overall, {jdgserver_name} system health
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+----
+
+=== Overall, {jdgserver_name} System Health
+Uses `kubectl` CLI tool to query the health status of {jdgserver_name} clusters and the {project_name} service in the specified namespace. This comprehensive check ensures that all components of the {project_name} deployment are operational and correctly configured within the Kubernetes environment.
```
```suggestion
Uses the `kubectl` CLI tool to query the health status of {jdgserver_name} clusters and the {project_name} service in the specified namespace. This comprehensive check ensures that all components of the {project_name} deployment are operational and correctly configured within the Kubernetes environment.
```

andymunro commented on docs/guides/high-availability/health-checks-multi-site.adoc:
```diff
@@ -0,0 +1,113 @@
+| jq 'reduce .[] as $item ([]; . + [keys[] | select($item[.] != "True")]) | if length == 0 then "HEALTHY" else "UNHEALTHY: " + (join(", ")) end'
+----
+
+=== {project_name} Readiness in Kubernetes
```
```suggestion
=== {project_name} readiness in Kubernetes
```

andymunro reviewed (changes requested):
Thanks, @kami619. Well written.  Just a few suggestions.

kami619 committed:
Apply suggestions from code review

Thanks @andymunro for the review. I have accepted your suggestions as they make sense to me.

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Kamesh Akella <kakella@redhat.com>

kami619 commented:
@ahus1 I think this should be all set now, after @andymunro's review and feedback.

ahus1 committed:
Avoiding active-active as in latest discussions

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>
